free-3.md

```diff
@@ -0,0 +1,230 @@
+debugging tools and more.
+
+Furthermore the details of some of the boilerplate may vary between trees
+written by a single author and definitely vary between trees written by
+different authors, making it just a bit harder to understand a tree written by
+someone else.
+
+I think that standardizing the way in which trees are written would help users
+understand what each other have written and would also save a lot of time spent
+writing boilerplate. This would also help users experiment with different ideas
+for tree algorithms more quickly.
+
+### Why a language?
+
+I think that this problem could be solved with an internal DSL.
+
+The DSL could provide a standardized way to specify the information being
+stored in the tree and the way algorithms which manipulate the structure of the
+tree or perform computations on it should work. Language seems like the natural
+way to do this standardization.
+
+A DSL would also have the ability to take most of the boilerplate out of the
+hands of the users, much in the say way that ScalaTest dramatically reduces the
+amount of code that needs to be written to do unit testing.
+
+One might reasonably suggest that creating some sort of abstract class which
+can be subclassed would be sufficient to solve this problem. The core of this
+observation is the question of whether such an implementation is really
+creating an internal DSL. I claim that it is. The result of people using such a
+super-class would be people writing different trees that all use a common
+vocabulary inherited from the super-class. I don't think this is the only way
+to solve the problem, but even if this was the solution I think it could be
+considered a DSL.
+
+### Why you?
+
+I find binary trees to be exceptionally interesting (dare I say magical?)
+data structures, and have spent a reasonable amount of time writing different
+types of trees in different ways. I think it would be really cool to have a
+standardized way of writing tree algorithms that gives you a whole bunch of
+debugging information and analysis tools right off the bat.
+
+### Domain
+
+Rapid Binary Tree Development
+
+### Interface (syntax)
+
+I imagine that one could do stuff vaguely like this:
+
+```
+Type T
+
+Node Contents:
+    data: T
+
+Node Procedure Insert(item: T):
+    if item < this.data:
+        if this.left exists:
+            this.left.Insert(item)
+        else:
+            this.left = new Node(item)
+    if item > this.data:
+        if this.right exists:
+            this.right.Insert(item)
+        else:
+            this.right = new Node(item)
+
+Tree Procedure Insert(item: T):
+    root.Insert(item)
+
+Tree Procedure Trolololol():
+    root.rotateRight()
+    root.rotateLeft()
+```
+
+The specific symbols used are not particularly important, but the idea is that
+standard binary tree ideas (left and right children, node contents, ordered
+types) are referred to in a standardize way.
```

Comment:
Most are pretty reasonable, but they can get somewhat exotic.

I think the real benefits are the tool chain and boilerplate.